ical-bot-matrix/matrix.go

```diff
@@ -0,0 +1,214 @@
+      // 	Algorithm:              id.AlgorithmMegolmV1,
+      //   // Recommended defaults
+      // 	RotationPeriodMillis:   7 * 24 * 60 * 60 *1000,
+      // 	RotationPeriodMessages: 100,
+      // }); err != nil {
+      //   slog.Warn("Could not upgrade room to encrypted", "roomID", evt.RoomID, "error", err)
+      // } else {
+      //   slog.Info("Successfully upgraded room to encrypted", "roomID", evt.RoomID)
+      // }
+		}
+
+		slog.Debug("Received a new message", "sender", evt.Sender.String(), "body", evt.Content.AsMessage().Body)
+		if evt.Sender != client.UserID {
+			resSend, err := client.SendText(ctx, evt.RoomID, "Yes, I heard you. Your message was "+evt.Content.AsMessage().Body)
+			if err != nil {
+				slog.Warn("Could not send message", "error", err)
+			}
+			slog.Debug("Sent message", "result", resSend)
+		}
+	})
+	syncer.OnEventType(event.StateMember, func(ctx context.Context, evt *event.Event) {
+		// StateKey is user's ID for member change events
+		if evt.GetStateKey() == client.UserID.String() {
+			membership := evt.Content.AsMember().Membership
+
+			switch membership {
+      case event.MembershipInvite:
+				resJoin, err := client.JoinRoom(ctx, evt.RoomID.String(), nil)
+				if err != nil {
+					slog.Error("Failed when attempting to join invited room", "roomID", evt.RoomID, "event", evt)
+				}
+
+				slog.Info("Successfully joined invited room", "result", resJoin)
+				sendMessage(client, ctx, evt.RoomID, "Hello, I'm the Ical Test Bot! It's nice to meet you :3")
+			default:
+				slog.Warn("Received unimplemented Membership change event", "type", membership, "event", evt)
+			}
+		}
+	})
+
+  cryptoStore := crypto.NewMemoryStore(nil)
+	// TODO: The docs are not entirely clear if the pickleKey is a secret or just a key
+	cryptoHelper, err := cryptohelper.NewCryptoHelper(client, []byte("awawawaaawa"), cryptoStore)
+	if err != nil {
+		slog.Error("Failed to setup CryptoHelper", "error", err)
+		os.Exit(1)
+	}
+	// TODO: Maybe setup cryptoHelper.Machine().log via slogzerolog
+	if err := cryptoHelper.Init(context.TODO()); err != nil {
+		slog.Error("Could not initialize cryptoHelper", "error", err)
+		os.Exit(1)
+	}
+	// De- and Encryption
+	client.Crypto = cryptoHelper
+	slog.Debug("Successfully initialized cryptohelper")
+
+	loginIcal()
+
+	var syncWait sync.WaitGroup
+	matrixSyncCtx, closeMatrixSyncContext := context.WithCancel(context.Background())
+
+	go func() {
+		// What does this do internally?
+		if err := client.SyncWithContext(matrixSyncCtx); err != nil {
+			if errors.Is(err, context.Canceled) {
+				slog.Error("Received cancel event. Shutting down...")
+			} else {
+				slog.Error("Received unknown Sync error", "error_type", err)
+			}
+		}
+	}()
+
+	syncWait.Add(1)
+	go func() {
+		defer syncWait.Done()
+		slog.Debug("Hiiii, I'm the Ical backend and I'm non-existent for meow")
+		fetchIcal()
+	}()
+
+	syncWait.Add(1)
+	go func() {
```

Comment:
You can also just place this in the main goroutine, there should be no need for this to be a seperate one, as it will just block main until it is ready to shut down. Running everything else in a Goroutine is fine

Reply:
I reworked and tested the go routines, do they look good to you?